src/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -528,42 +523,69 @@ private static string FormatBigIntegerToHexString(BigInteger value, char format,
                 // [07..00] drop the high 0 as the two's complement positive number remains clear
                 bool clearHighF = false;
                 byte head = bits[cur];
+
                 if (head > 0xF7)
                 {
                     head -= 0xF0;
                     clearHighF = true;
                 }
+
                 if (head < 0x08 || clearHighF)
                 {
                     // {0xF8-0xFF} print as {8-F}
                     // {0x00-0x07} print as {0-7}
-                    fmt = string.Format(CultureInfo.InvariantCulture, "{0}1", format);
-                    sb.Append(head.ToString(fmt, info));
+                    sb.Append(head < 10 ?
+                        (char)(head + '0') :
+                        format == 'X' ? (char)((head & 0xF) - 10 + 'A') : (char)((head & 0xF) - 10 + 'a'));
                     cur--;
                 }
             }
+
             if (cur > -1)
             {
-                fmt = string.Format(CultureInfo.InvariantCulture, "{0}2", format);
-                while (cur > -1)
+                Span<char> chars = sb.AppendSpan((cur + 1) * 2);
+                int charsPos = 0;
+                if (format == 'x')
                 {
-                    sb.Append(bits[cur--].ToString(fmt, info));
+                    while (cur > -1)
+                    {
+                        byte b = bits[cur--];
+                        chars[charsPos++] = GetLowerCaseHexValue(b >> 4);
+                        chars[charsPos++] = GetLowerCaseHexValue(b & 0xF);
```

Comment:
Instead of two functions and splitting the format; would it make more sense to have a local char for 'A' vs 'a'?

Reply:
> Instead of two functions and splitting the format; would it make more sense to have a local char for 'A' vs 'a'?

Can you elaborate?

Reply:
```C#
var sb = new ValueStringBuilder(stackSpace);
char alphaBase = format == 'x' ? 'a' : 'A';

...
                 if (head < 0x08 || clearHighF)
                 {
                     sb.Append(head < 10 ? (char)(head + '0') : (char)((head-10) + alphaBase));
                 }

            ...

            chars[charPos++] = GetHexValue(b >> 4, alphaBase);
            ...

private static char GetHexValue(int i, char alphaBase) => i < 10 ? (char)(i + '0') : (char)(i - 10 + alphaBase);
```

Depending on how optimized string or span indexing is, it might be better to dump the ternary expression math and just index into two strings.

```C#
private const string LowerHex = "0123456789abcdef";
private const string UpperHex = "0123456789ABCDEF";

...
var sb = new ValueStringBuilder(stackSpace);
string basis = format == 'x' ? LowerHex : UpperHex;

chars[charsPos++] = basis[b >> 4];
chars[charsPos++] = basis[b & 0xF];
```

Reply:
I like that string version.  Will switch.